### ISSUE TITLE
Release Twisted 14.0

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7039]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7039|
|Type|task|
|Created|2014-03-19 09:43:37Z|
|Branch|https://github.com/twisted/twisted/tree/no-conflict-14.0-7039|

Better get this out before ~~PyCon~~ the heat death of the universe!

<details><summary>Searchable metadata</summary>

```
trac-id__7039 7039
type__task task
reporter__hawkowl hawkowl
priority__highest highest
milestone__Twisted_14_0_0 Twisted-14.0.0
branch__branches_no_conflict_14_0_7039 branches/no-conflict-14.0-7039
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__ 
time__1395222217000000 1395222217000000
changetime__1399893318000000 1399893318000000
version__None None
owner__hawkowl hawkowl
cc__radix cc__hynek
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

[#6926](https://github.com/twisted/twisted/issues/6926) really needs to be a blocker for this.  There are (obviously, if you've spoken to Hynek) other things that would be nice to land, but that's a documentation regression that's in the milestone.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

Oh, crud - I totally didn't see that one. :(

I'll work on [#6926](https://github.com/twisted/twisted/issues/6926) tonight.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

> [#6926](https://github.com/twisted/twisted/issues/6926) really needs to be a blocker for this. There are (obviously, if you've spoken to Hynek) other things that would be nice to land, but that's a documentation regression that's in the milestone.

Then it should be marked as a regression.  The release process has a step that involves checking for regressions but this doesn't help if the regression tickets are filed as "enhancements".

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

I'm wanting to push forward with 14.0 and add the fixes for the warnings into 14.0.1.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

What about [#7097](https://github.com/twisted/twisted/issues/7097)?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl commented|
|-|-|

[#7097](https://github.com/twisted/twisted/issues/7097) will go in 14.0.1, which can have the 14.0.1pre1 prerelease with the fix right after 14.0 release.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

Why?  [#7097](https://github.com/twisted/twisted/issues/7097) is fixed (I spent time fixing it, by the way, so it would go into 14.0.0).  It fixes a problem that is being introduced by 14.0.0 as-is.  It was marked as a regression.

It really belongs in 14.0.0.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl commented|
|-|-|

Well, looks like I'm rolling pre6, then.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

I didn't notice the fact that [[#7097](https://github.com/twisted/twisted/issues/7097)](https://github.com/twisted/twisted/issues/7097) was tagged as a regression, but the ticket doesn't mention anywhere on it that you did the work to get it into 14.0.  In the future, please add tickets to the milestone so we have some visibility into which things people are proposing for addition to the release post-prerelease.  I've done so now.

We may want to revisit the way the release process works, generally.  Following the existing release process and schedule schedule, we're now going to be delaying the final release of a super important security fix for a week because of some whitespace in a warning message.  At the very least, the intent to include #7097 in 14.0 should have been surfaced as quickly as possible as soon as it landed, so that we immediately issued a pre6 and minimized the additional delay.  As I understand it here nobody else was aware that you had tagged it as a regression.  (I also think that classifying an error in the _presentation_ of new information as a "regression" is a bit of a stretch.)

Nevertheless let's follow the process we've got for now; maybe we can change it next time.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Sorry, I thought it had already been agreed this fix would go into a 14.0 pre-release: <https://twistedmatrix.com/pipermail/twisted-python/2014-April/028221.html>.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_15):
> Sorry, I thought it had already been agreed this fix would go into a 14.0 pre-release: <https://twistedmatrix.com/pipermail/twisted-python/2014-April/028221.html>.

Fair enough, that's a reasonable interpretation of that comment.

Still, in the future, better to be more definitive at each step; when the bug is created (add it to the milestone) when it's completed (tell the release manager on the mailing list).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

> Still, in the future, better to be more definitive at each step; when the bug is created (add it to the milestone) when it's completed (tell the release manager on the mailing list).

Sure.  Is this written down somewhere better?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

I reviewed [the release process page](https://twistedmatrix.com/trac/wiki/ReleaseProcess) and ... it ... sort of says some things about this but not really in a useful form for contributors.  It did draw my attention to the fact that [we have a couple of other open regressions](https://twistedmatrix.com/trac/query?status=new&status=assigned&status=reopened&type=regression&order=priority) which is sad.

For now, I'll add something to [the review process page](https://twistedmatrix.com/trac/wiki/ReviewProcess) in the "merge a change to trunk" section for things about notifying the release manager.  We should bug hawkowl to figure out where to put the stuff about what to do when you notice something _should_ block the release.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

Okay I said a few things about announcing release regressions more broadly too.  [[ReviewProcess](https://twistedmatrix.com/trac/wiki/ReviewProcess)?action=diff&version=62&old_version=61 Here are the changes.]

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_22"></a>|@itamarst set owner to @hawkowl|
|-|-|

I'm looking at https://github.com/twisted/twisted/compare/trunk...releases/release-14.0.0-7039-3, rather than -4 based on IRC discussion.

1. It still says 14.0.0pre5 - not sure if automated process fixes this or what. Maybe the automated tool does actually fix this when you run it, in which case it's fine. But maybe I should be reviewing the final output of that tool?
2. Need new quote-of-the-release, I think? I feel the one about TLS PFS is apropos.
3. The diff I'm seeing includes many changes to code and examples and what not, which seems bad - merging that will presumably undo ongoing work in trunk.

... possibly we need to restart the release process from trunk :(

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph commented|
|-|-|

HawkOwl started a ... somewhat premature, I guess, since this ticket isn't closed yet ... post-mortem discussion on the mailing list: https://twistedmatrix.com/pipermail/twisted-python/2014-May/028336.html

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_25"></a>|@hawkowl removed owner|
|-|-|

The branch is now in 14.0 final state. Buildbots are going green except for a known bug on a builder, which will be fixed in 14.1.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_26"></a>|@hawkowl commented|
|-|-|

Elaboration:

https://buildbot.twistedmatrix.com/builders/rhel6-x86_64-py2.6 is failing, due to a bug in 14.0's UDP multicast code. Tom Prince committed a fix for this ([#7159](https://github.com/twisted/twisted/issues/7159)) in current trunk, which catches the error, with a comment stating that is a flaw of RHEL6 and not Twisted. Furthermore, this issue appears in 13.2 ( https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/releases/release-13.2.0-6795 ) so it's not a regression, but a bug.

This release does not contain [#7097](https://github.com/twisted/twisted/issues/7097) due to code churn making the merge have potentially more problems than it solves. This fix will be in 14.1, of which I will issue a prerelease as soon as 14.0 closes, to make up for its non-inclusion in 14.0. The RHEL fix will also be in 14.1, which means that we will properly support it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

Since the changes in the release branch will necessarily conflict with the changes in trunk I suggest that when this is being merged into trunk the working copy with conflicts resolved be reviewed as well (*before* the merge is committed to trunk).

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_28"></a>|@hawkowl commented|
|-|-|

Yes, that seems like a very good idea, and I'll do that once it's ready for merging.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_29"></a>|@hynek commented|
|-|-|

Since the discussion happened unfortunately on 10 different communication channels, let me re-iterate if I got everything right:

 - the widely tested pre5 is supposed to be the final without further changes (cf. jp’s typo concern)
 - there are no open questions left (cf. hundreds of lines in IRC)
 - merging back is going to be a completely separate topic (cf. itamar’s concern on the ML)
 - no food is sustainable

Is that correct?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_30"></a>|@hawkowl commented|
|-|-|

RE: hynek:

- yes

- yes

- merging back will be a minor issue, only two small conflicts, as noted by glyph on the ML and me testing it

- yes

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_31"></a>|@hynek commented|
|-|-|

I went through the diff and noticed that in the NEWS files, each section has changes that has been added in later pres added as a new section.  In the case of Twisted Web it’s especially hilarious:


```
Twisted Web 14.0.0 (2014-05-08)
===============================

No significant changes have been made for this release.


Twisted Web 14.0.0 (2014-05-08)
===============================

Bugfixes
--------
 - When requesting an HTTPS URL with twisted.web.client.Agent, the
   hostname of the presented certificate will be checked against the
   requested hostname; mismatches will now result in an error rather
   than a man-in-the-middle opportunity for attackers. This may break
   existing code that incorrectly depended on insecure behavior, but
   such code was erroneous and should be updated. (#4888)


Twisted Web 14.0.0 (2014-05-08)
===============================

Features
--------
 - twisted.web.http.proxiedLogFormatter can now be used with
   twisted.web.http.HTTPFactory (and subclasses) to record X
   -Forwarded-For values to the access log when the HTTP server is
   deployed behind a reverse proxy. (#1468)
 - twisted.web.client.Agent now uses
   twisted.internet.ssl.CertificateOptions for SSL/TLS and benefits
   from its continuous improvements. (#6893)

Bugfixes
--------
 - twisted.web.client.Agent now correctly manage flow-control on
   pooled connections, and therefore requests will no longer hang
   sometimes when deliverBody is not called synchronously within the
   callback on Request. (#6751)
 - twisted.web.client.Agent now verifies that the provided server
   certificate in a TLS connection is trusted by the platform. Until
   #4888 is completed there are still known MITM attack vectors
   against Agent. Applications which were relying on the absence of
   certificate checking will now receive an SSL error. (#7042)

Other
-----
 - #5004, #6881, #6956
```


I presume that’s not on purpose?  The news files should definitely be collapsed to have only one `14.0.0 (2014-05-08)` section.

Other than that it looks good to me but please re-submit for review after adressing this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_32"></a>|@glyph commented|
|-|-|

The "Until [[#4888](https://github.com/twisted/twisted/issues/4888)](https://github.com/twisted/twisted/issues/4888)..." part of the Twisted Web section is also kind of nonsensical note, given that we just finished telling the user about how #4888 was already fixed.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_33"></a>|@hawkowl commented|
|-|-|

Well, now [#4888](https://github.com/twisted/twisted/issues/4888) is after it, so it makes logical sense!

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun set owner to @hawkowl|
|-|-|

Thanks.

  1. [#4500](https://github.com/twisted/twisted/issues/4500) seems like a documentation ticket but it is in the features section of the news files.
  1. Since both [[#7042](https://github.com/twisted/twisted/issues/7042)](https://github.com/twisted/twisted/issues/7042) and [#4888](https://github.com/twisted/twisted/issues/4888) are being resolved at the same time, the news fragment for #7042 doesn't really make sense as written.  The "Until ..." sentence should just be deleted from the news files.
  1. The news items for [#11556](https://github.com/twisted/twisted/issues/11556) are inconsistent in the two news files they appear in.  Were any other items manually edited in only one of the two news files?
  1. There are, of course, merge conflicts, the resolution of which I can't review yet.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_35"></a>|@hawkowl removed owner|
|-|-|

1, 2 and 3 are fixed. 4 will be fixed once the release is cut, to avoid changing the code - unless I'm missing something.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_36"></a>|@hawkowl commented|
|-|-|

(In [42630]) Branching to no-conflict-14.0-7039.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_37"></a>|@hawkowl commented|
|-|-|

Fixed 4, with the buildbot results being here: https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/no-conflict-14.0-7039

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_38"></a>|@hynek commented|
|-|-|

As far as I’m concerned, this is ready to go.  I’ll let the review keyword on until tomorrow UTC-morning to give others a chance to veto but would put this ticket out of its misery (8 weeks :|) then.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_39"></a>|@hynek set owner to @hawkowl|
|-|-|

I noticed there’s one stray whitespace before the `(#4888)`, you may optionally want to fix that.

That aside:

RELEASE THE KRAKEN

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_40"></a>|@hawkowl set status to `closed`|
|-|-|

(In [42638]) Merge no-conflict-14.0-7039: Release Twisted 14.0.0

Author: hawkowl
Reviewers: exarkun, itamar, hynek
Fixes: [#7039](https://github.com/twisted/twisted/issues/7039)

Release Twisted 14.0.0
